### DISCUSSION TITLE
如何优雅的互相调用http服务？

### DISCUSSION THREAD

**Discussion by xmsz** [midwayjs/midway#1423] [Q&A]
背景
- 我有几个服务，独立的项目，假设userService和articleService，
- 我想在userService中调用articleService（微服务调用）

现在
- 普通情况通过http服务，.get(articleService/api/xxx)。这样的话，如果articleService的controller改了，那同步就断开了，而且也没有类型的提示
- 然后虽然有grpc微服务调用，但是变得更麻烦

需求
调用普通http项目服务的时候，和调用微服务一样

```

const res = await articleService.call("getList")

```
同时也支持类型提示

---------------------------

我想手动实现，但是没找到好的思路，大概就是

拿到controller的信息，生成
```
interface IService  {
 user: {
   getUserInfo(userId: number): Promise<{name: string}>
 }
}
const service:IService = {
 user : {
  getUserInfo: (userId) => {
      return request.get('/api/user/info')
  }
 }
}
```

**Comment by czy88840616** [comment]
你可以把服务抽成 npm 包，然后直接引用。。

**Reply by xmsz** [reply]
还是需要通过服务调用 因为是单独部署

**Comment by czy88840616** [comment]
http 就走标准的请求呗，axios 啥的，grpc 也是一样的。

**Reply by xmsz** [reply]
没错 然后我想自动生成和grpc调用一样的，类型和调用方法
这样不用再单独维护一份

**Reply by czy88840616** [reply]
这个其实一直是比较困难的点，类似我们的一体化方案是前端调用后端，还有一种是类似swagger之后，利用swagger信息生成通用的sdk调用方案。
